Raven.Client.Lightweight/Changes/RemoteChangesClientBase.cs

```diff
@@ -201,83 +213,133 @@ private bool DisposeHeartbeatTimer()
             }
         }
 
-        private Task lastSendTask;
-
-        protected Task Send(string command, string value)
+        async Task Worker()
         {
-            lock (this)
+            while (tokenSource.IsCancellationRequested == false)
             {
-                logger.Info("Sending command {0} - {1} to {2} with id {3}", command, value, url, id);
-                var sendTask = lastSendTask;
-                if (sendTask != null)
+                try
                 {
-                    return sendTask.ContinueWith(_ =>
-                    {
-                        Send(command, value);
-                    });
+                    await syncSource.Task.ConfigureAwait(false);
+                    syncSource = TaskCompletionSourceFactory.Create<bool>();
                 }
-
-                try
+                catch (TaskCanceledException)
                 {
-                    var sendUrl = url + "/changes/config?id=" + id + "&command=" + command;
-                    if (string.IsNullOrEmpty(value) == false)
-                        sendUrl += "&value=" + Uri.EscapeUriString(value);
+                    // Swallowing the task cancellation in case we are stopping work.
+                }
 
-                    var requestParams = new CreateHttpJsonRequestParams(null, sendUrl, HttpMethods.Get, credentials, Conventions)
-                    {
-                        AvoidCachingRequest = true
-                    };
-                    var request = jsonRequestFactory.CreateHttpJsonRequest(requestParams);
-                    lastSendTask = request.ExecuteRequestAsync().ObserveException();
+                await WorkUntilEmpty().ConfigureAwait(false);
+            }
 
-                    return lastSendTask.ContinueWith(task =>
-                    {
-                        lastSendTask = null;
-                        request.Dispose();
-                    });
-                }
-                catch (Exception e)
+            // drain
+            await WorkUntilEmpty().ConfigureAwait(false);
+        }
+
+        private async Task WorkUntilEmpty()
+        {
+            Work work;
+            while (workQueue.TryDequeue(out work))
+            {
+                using (work)
                 {
-                    return new CompletedTask(e).Task.ObserveException();
+                    try
+                    {
+                        await work.SendTask.ConfigureAwait(false);
+                        work.DoneTask.TrySetResult(true);
+                    }
+                    catch (Exception e)
+                    {
+                        work.DoneTask.TrySetException(e);
```

Comment:
This line could also be removed if the previous behavior of swalling exceptions is of interest/desired. Then the TrySet would need to move into a finally

Reply:
The issue is the breaking change, I think

Reply:
Then again, I wonder how many people actually do wait on this

Reply:
That's why I mentioned it. I can keep swallowing the exception